site/docs/expressions/Extension_Functions/index.md

```diff
@@ -0,0 +1,219 @@
+
+!!! note
+
+    Constant value arguments effectively always act like scalars, even for function types that take vectors as input. For example, an aggregate function like `count_matching(regex: const string, values: string) -> i64` would always only need to compile the passed regular expression once, while `count_matching(regex: string, values: string) -> i64` is expected to work even if the regex varies from record to record.
+
+### Type Arguments
+
+Type arguments are arguments that are used only to inform the evaluation and/or type derivation of the function. For example, you might have a function which is `truncate(type DECIMAL<P0,S0>, DECIMAL<P1, S1>, i32)`. This function declares two value arguments and a type argument. The difference between them is that the type argument has no value at runtime, while the value arguments do.
+
+| Property | Description                                     | Required |
+|----------|-------------------------------------------------|----------|
+| Metadata | See [common metadata](#common-metadata).        | Optional |
+| Pattern  | A [data type pattern](#data-type-patterns).     | Required |
+
+### Required Enumeration Arguments
+
+Required enumerations are arguments that support a fixed set of declared values as constant arguments. These arguments must be specified as part of an expression. While these could also have been implemented as constant string value arguments, they are formally included to improve validation/contextual help/etc. for frontend processors and IDEs. An example might use might be `extract({DAY, YEAR, MONTH}, date) -> i32`. In this example, a producer must specify a type of date part to extract. Note that the value of a required enumeration cannot be used in return type derivation.
+
+| Property | Description                                     | Required |
+|----------|-------------------------------------------------|----------|
+| Metadata | See [common metadata](#common-metadata).        | Optional |
+| Options  | List of valid string options for this argument. | Required |
+
+### Optional Enumeration Arguments
+
+Optional enumeration arguments are similar to required enumeration arguments, but are more focused on supporting alternative behaviors, usually for corner cases. An optional enumeration always includes an "unspecified" default option that can be bound based on the capabilities of the plan consumer. When a plan does not specify a behavior, the consumer is expected to resolve the option based on the first option the system can match. An example use case might be `OVERFLOW_BEHAVIOR: {OVERFLOW, SATURATE, ERROR}`. If unspecified, an engine would use the first of these that it implements. If specified, the engine would be expected to behave as specified or fail. Note that the value of an optional enumeration cannot be used in type derivation.
```

Comment:
> What does error mean here in terms of the function's return type?

It's a common option that's used all over the place. Will be documented by one of my open PRs. https://github.com/jvanstraten/substrait/blob/c5c057deaff428d94ebb006251540463c245ff5a/site/docs/extensions/core_extensions/common_options.md?plain=1#L5-L14

> Does this mean that function return types can't change based on the value of an optional enum arg?

Yes, but that's actually sensible, because the point of optional enumeration arguments is to allow the producer to hand control over some implementation details over the the consumer. It's used for things that are *technically* different behavior across pretty much all query engines, but users rarely care because it's about corner cases that should never come up in the first place or you were screwed from the get-go (like overflow, floating point rounding modes, etc.). If the consumer can choose which option it wants, and the return type depends on the option, no one except the consumer would be able to do type propagation.